pkg/admissioncontroller/webhooks/admission/auditpolicy/admission.go

```diff
@@ -0,0 +1,290 @@
+	)
+	if request.Operation != admissionv1.Create && request.Operation != admissionv1.Update {
+		return acadmission.Allowed("operation is not Create or Update")
+	}
+
+	if request.SubResource != "" {
+		return acadmission.Allowed("subresources are not handled")
+	}
+
+	if err := h.decoder.Decode(request, shoot); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if !hasAuditPolicy(shoot) {
+		return acadmission.Allowed("shoot resource is not specifying any audit policy")
+	}
+	cmRef := shoot.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy.ConfigMapRef
+
+	auditPolicyCm := &corev1.ConfigMap{}
+	if err := h.apiReader.Get(ctx, kutil.Key(shoot.Namespace, cmRef.Name), auditPolicyCm); err != nil {
+		if apierrors.IsNotFound(err) {
+			return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("referenced audit policy does not exist: namespace: %s, name: %s", shoot.Namespace, cmRef.Name))
+		}
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("could not retrieve config map: %s", err))
+	}
+
+	if shoot.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy.ConfigMapRef.ResourceVersion == auditPolicyCm.ResourceVersion {
+		return acadmission.Allowed("no change detected in referenced configmap holding audit policy")
+	}
+
+	auditPolicy, err := getAuditPolicy(auditPolicyCm)
+	if err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("error getting auditlog policy from ConfigMap %s/%s: %w", shoot.Namespace, cmRef.Name, err))
+	}
+
+	schemaVersion, errCode, err := validateAuditPolicySemantics(auditPolicy)
+	if err != nil {
+		return admission.Errored(errCode, err)
+	}
+
+	// Validate audit policy schema version against k8s version of shoot
+	if isValidVersion, err := IsValidAuditPolicyVersion(shoot.Spec.Kubernetes.Version, schemaVersion); err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	} else if !isValidVersion {
+		err := fmt.Errorf("your shoot cluster kubernetes version %q is not compatible with audit policy version %q", shoot.Spec.Kubernetes.Version, schemaVersion.GroupVersion().String())
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	}
+
+	// If the resource Version in the audit policy cm changed it also needs to be adapted in the configMapRef
+	return admission.Patched(
+		"referenced audit policy is valid",
+		jsonpatch.NewOperation("replace", "/spec/kubernetes/kubeAPIServer/auditConfig/auditPolicy/configMapRef/resourceVersion", auditPolicyCm.ResourceVersion),
+	)
+}
+
+func (h *handler) admitConfigMap(ctx context.Context, request admission.Request) admission.Response {
+	var (
+		oldCm = &corev1.ConfigMap{}
+		cm    = &corev1.ConfigMap{}
+	)
+
+	if err := h.getOldObject(request, oldCm); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if request.Operation != admissionv1.Update {
+		return acadmission.Allowed("operation is not update")
+	}
```

Comment:
Shouldn't this be moved above before trying to get the old object?

Reply:
definitely!